JetBrains: Add storybook for Chromatic tests

vdavid opened the pull request:
Fixes https://github.com/sourcegraph/sourcegraph/issues/37681

I made webpack work with the project (needed some polyfills) and added a story.

Also:
 - Bug fix: The initial search was broken in the standalone version. Now it correctly triggers.
 - Bug fix: Path matches didn't send the "commit" field so the "Copy URL" feature didn't work for remote files in path matches. (I've found this out by accident during testing)
 - Extracted the mocked "callJava" function: it's reusable now (I've reused it) and more elegant too.

This is how it looks in both modes:

<div>
<img width="1085" alt="CleanShot 2022-07-12 at 19 49 45@2x" src="https://user-images.githubusercontent.com/2552265/178559799-c85281bf-992c-4ff3-a66a-5732d0bf937e.png">
<img width="1090" alt="CleanShot 2022-07-12 at 19 52 11@2x" src="https://user-images.githubusercontent.com/2552265/178559918-a8153eb8-ac5d-45ad-8cad-632ee1110e5c.png">
<div>

Caveat:
 - To avoid a flaky test and make sure that the search finishes by the time Chromatic creates its snapshot, I set up a 10 second delay after page loading. I imagine that Chromatic works in a heavily multithreaded way (as we're generating hundreds of pages) so this should be non-blocking. I've tried to find a more dynamic solution, but haven't found anything in the docs and on Google, so I stayed with the fixed delay.

## Test plan

I've tested all of these:
- It works with `yarn storybook:jetbrains`
- It works with `yarn storybook`
- It works with both dark and light modes correctly
- It still works with the standalone version
- It still works in the IDE

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-dv-jetbrains-chromatic.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-tmnegtqwet.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

vdavid commented:
@taras-yemets @philipp-spiess any ideas how to make the CSS work?
(Can be tested by checking out the branch and running `yarn storybook` in the sourcegraph root.)

philipp-spiess commented:
@vdavid I was so free and pushed something onto your branch that should help with that. We have to manually add the global stylesheets. Take a look at https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Fsourcegraph.git&branch=HEAD&file=client%2Fbranded%2Fsrc%2Fcomponents%2FBrandedStory.tsx&editor=VSCode&version=2.2.7&start_row=0&start_col=12&end_row=0&end_col=12&utm_campaign=vscode-extension&utm_medium=direct_traffic&utm_source=vscode-extension&utm_content=vsce-commands where we do this for other storybook tests.

I also added the global `theme` and `theme-light` config but I think we ideally want to run [this code](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/jetbrains/webview/src/search/index.tsx?L95-129&subtree=true) with one of the theme snapshots to also have the custom CSS vars defined.

philipp-spiess commented on client/jetbrains/package.json:
```diff
@@ -13,13 +13,14 @@
     "lint:css": "stylelint 'webview/**/*.scss'",
-    "typecheck": "tsc -b"
+    "standalone": "ts-node standalone/src/server.ts",
+    "storybook": "STORIES_GLOB=client/jetbrains/webview/src/**/*.story.tsx yarn workspace @sourcegraph/storybook run start",
```
The re-ordering makes it really hard to see what the change is, Is it only the addition of this command? If so then LGTM

philipp-spiess reviewed (commented):
> I imagine that Chromatic works in a heavily multithreaded way (as we're generating hundreds of pages) so this should be non-blocking. 

Any way we can be sure about this? How do other snapshots handle this? And even if these tests are parallelized, it could still be part of the slowest path in the graph and thus delay our whole run by 10 seconds. 

I also noticed that we currently do not apply the JetBrains specific CSS vars. Do you think it is okay to not include them or should we add them based on the theming snapshots that we already have bundled? 

Also: Does it make sense to include other result types in this static snapshot so that we have a higher coverage?

philipp-spiess reviewed (approved):
Stamping it regardless so you can work out answers to my above questions and ship what you find useful without having to wait for another review :)

vdavid commented on client/jetbrains/webview/src/bridge-mock/call-java-mock.ts:
```diff
@@ -1,6 +1,6 @@
 import { decode } from 'js-base64'
 
-import { SearchPatternType } from '@sourcegraph/shared/out/src/graphql-operations'
+import { SearchPatternType } from '@sourcegraph/search'
```
Ah, thanks, @philipp-spiess, apparently that was me back in May.

philipp-spiess replied:
Haha yeah the build was failing because of this 😬  No worries!

vdavid replied:
I'm not sure how that's possible. It's a code line that I moved here from `bridge-mock/index.ts`.

See? It's in `main` now: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/jetbrains/webview/src/bridge-mock/index.ts?L3%3A25=

vdavid commented:
> > I imagine that Chromatic works in a heavily multithreaded way (as we're generating hundreds of pages) so this should be non-blocking.
> Any way we can be sure about this? How do other snapshots handle this? And even if these tests are parallelized, it could still be part of the slowest path in the graph and thus delay our whole run by 10 seconds.

I tried looking for examples, but we do no network requests in any of our storybook pages where Chromatic is enabled.
So I removed the App test, and added two individual tests, one for the search box, one for the result list.

> I also noticed that we currently do not apply the JetBrains specific CSS vars. Do you think it is okay to not include them or should we add them based on the theming snapshots that we already have bundled?

I saw you added this yesterday, thanks for that!

> Also: Does it make sense to include other result types in this static snapshot so that we have a higher coverage?

It does. I've added them to the new tests. This is how they look:

<img width="931" alt="CleanShot 2022-07-15 at 11 27 29@2x" src="https://user-images.githubusercontent.com/2552265/179195569-05927560-94a0-4abc-9ed2-dc07728854d9.png">

<img width="942" alt="CleanShot 2022-07-15 at 11 27 43@2x" src="https://user-images.githubusercontent.com/2552265/179195609-9b2cbbcb-9441-47f6-9336-01b4d618905e.png">

@philipp-spiess WDYT?

philipp-spiess commented:
@vdavid 🚀  Awesome!!
